clodius/tiles/fasta.py

```diff
@@ -0,0 +1,260 @@
+
+
+
+def natsort_key(s, _NS_REGEX=re.compile(r"(\d+)", re.U)):
+    return tuple([int(x) if x.isdigit() else x for x in _NS_REGEX.split(s) if x])
+
+
+def natcmp(x, y):
+    if x.find("_") >= 0:
+        x_parts = x.split("_")
+        if y.find("_") >= 0:
+            # chr_1 vs chr_2
+            y_parts = y.split("_")
+
+            return natcmp(x_parts[1], y_parts[1])
+        else:
+            # chr_1 vs chr1
+            # chr1 comes first
+            return 1
+    if y.find("_") >= 0:
+        # chr1 vs chr_1
+        # y comes second
+        return -1
+
+    _NS_REGEX = re.compile(r"(\d+)", re.U)
+    x_parts = tuple([int(a) if a.isdigit() else a for a in _NS_REGEX.split(x) if a])
+    y_parts = tuple([int(a) if a.isdigit() else a for a in _NS_REGEX.split(y) if a])
+
+    # order of these parameters is purposefully reverse how they should be
+    # ordered
+    for key in ["m", "y", "x"]:
+        if key in y.lower():
+            return -1
+        if key in x.lower():
+            return 1
+
+    try:
+        if x_parts < y_parts:
+            return -1
+        elif y_parts > x_parts:
+            return 1
+        else:
+            return 0
+    except TypeError:
+        return 1
+
+
+def natsorted(iterable):
+    return sorted(iterable, key=ft.cmp_to_key(natcmp))
+
+
+def get_chromsizes(fapath):
+    with Fasta(fapath, one_based_attributes=False) as fa:
+        chromsizes = dict((seq, len(fa.records[seq])) for seq in fa.keys())
+        chromosomes = natsorted(fa.keys())
+    return pd.Series(chromsizes)[chromosomes]
+
+
+def tileset_info(fapath, chromsizes=None):
+    """
+    Get the tileset info for a FASTA file
+
+    Parameters
+    ----------
+    fapath: string
+        The path to the FASTA file from which to retrieve data
+    chromsizes: [[chrom, size],...]
+        A list of chromosome sizes associated with this tileset.
+        Typically passed in to specify in what order data from
+        the FASTA should be returned.
+
+    Returns
+    -------
+    tileset_info: {'min_pos': [],
+                    'max_pos': [],
+                    'tile_size': 1024,
+                    'max_zoom': 7
+                    }
+    """
+    TILE_SIZE = 1024
+
+    if chromsizes is None:
+        chromsizes = get_chromsizes(fapath)
+        chromsizes_list = []
+
+        for chrom, size in chromsizes.iteritems():
+            chromsizes_list += [[chrom, int(size)]]
+    else:
+        chromsizes_list = chromsizes
+
+    min_tile_cover = np.ceil(sum([int(c[1]) for c in chromsizes_list]) / TILE_SIZE)
+    max_zoom = int(np.ceil(np.log2(min_tile_cover)))
+
+    tileset_info = {
+        "min_pos": [0],
+        "max_pos": [TILE_SIZE * 2 ** max_zoom],
+        "max_width": TILE_SIZE * 2 ** max_zoom,
+        "tile_size": TILE_SIZE,
+        "max_zoom": max_zoom,
+        "chromsizes": chromsizes_list
+    }
+    return tileset_info
+
+
+def abs2genomic(chromsizes, start_pos, end_pos):
```

Comment:
OK, we've definitely set a bad precedent by duplicating this function in clodius/tiles/bam.py, clodius.tiles.bigwig.py, and clodius.tiles.multivec.py but if you're feeling generous and want to refactor it into clodius.tiles.utils, it would be greatly appreciated. If not, no worries, we'll do it another time.

Reply:
Will do.